src/openforms/variables/rendering/nodes.py

```diff
@@ -0,0 +1,90 @@
+from dataclasses import dataclass
+from typing import Any, Iterator, Union
+
+from django.utils.translation import gettext_lazy as _
+
+from openforms.forms.constants import FormVariableSources
+from openforms.submissions.models import Submission, SubmissionValueVariable
+from openforms.submissions.rendering import RenderModes
+from openforms.submissions.rendering.base import Node
+
+
+@dataclass
+class VariablesNode(Node):
+    """
+    Render node for the user defined variables related to a form.
+
+    This node is only 'visible' for certain render modes (cli and registration), but it is not rendered if there are
+    user defined variables related to the form.
+
+    Rendering this node outputs the name of the step within the form.
+    """
+
+    submission: Submission
+
+    _variables = None
+
+    @property
+    def is_visible(self) -> bool:
+        if self.mode not in [RenderModes.cli, RenderModes.registration]:
+            return False
+        return len(self.variables) > 0
+
+    @property
+    def variables(self):
+        if not self._variables:
+            self._variables = self.submission.submissionvaluevariable_set.filter(
+                form_variable__source=FormVariableSources.user_defined
+            ).select_related("form_variable")
+        return self._variables
+
+    def render(self) -> str:
+        if not self.is_visible:
+            return ""
+        return _("Variables")
+
+    def get_children(self) -> Iterator["Node"]:
+        for variable in self.variables:
+            node = SubmissionValueVariableNode(
+                renderer=self.renderer, variable=variable
+            )
+            if node.is_visible:
+                yield node
+
+
+@dataclass
+class SubmissionValueVariableNode(Node):
+    variable: SubmissionValueVariable
+    depth: int = 0
+    is_layout: bool = False
+
+    @property
+    def is_visible(self) -> bool:
+        if self.mode in [RenderModes.export, RenderModes.registration, RenderModes.cli]:
+            return True
+        return False
+
+    @property
+    def value(self) -> Any:
+        return self.variable.value
+
+    def get_children(self) -> Iterator["Node"]:
+        return iter([])
+
+    @property
+    def label(self) -> str:
+        if self.mode == RenderModes.export:
+            return self.variable.key
+        return self.variable.form_variable.name
+
+    @property
+    def display_value(self) -> Union[str, Any]:
+        # in export mode, expose the raw datatype
+        if self.mode == RenderModes.export:
+            return self.variable.value
+
+        # TODO add a registry of formatters like for Formio ComponentNodes
```

Comment:
I wasn't sure if I should do this, since we are only rendering for registration, cli and export (not for submission report pdf, confirmation email or summary page) :thinking:

Reply:
let's not do it, but add a comment that this would be needed if we'll render in different modes.